core/include/atomic.h

```diff
@@ -35,9 +46,101 @@
  */
 unsigned int atomic_set_return(unsigned int *val, unsigned int set);
 
+/**
+ * @brief Atomic Compare and Swap
+ *
+ * Updates the value in var iff the current value of var is equal to old.
+ *
+ * @param[inout] var  Atomic variable to update
+ * @param[in]    old  The old value to compare against
+ * @param[in]    now  The new value to write to var
+ *
+ * @return 1 if the write completed successfully
+ * @return 0 if the write failed.
+ */
+int atomic_cas(atomic_int_t *var, int old, int now);
+
+/**
+ * @brief Increment a counter variable by one atomically and return the old value.
+ *
+ * @param[inout]  var   Pointer to a counter variable.
+ *
+ * @return The value of *val* before the increment.
+ */
+static inline int atomic_inc(atomic_int_t *var)
+{
+    int old;
+
+    do {
+        old = var->value;
+    } while (!atomic_cas(var, old, old + 1));
+
+    return old;
+}
+
+/**
+ * @brief Decrement a counter variable by one atomically and return the old value.
+ *
+ * @param[inout]  var   Pointer to a counter variable.
+ *
+ * @return The value of *val* before the decrement.
+ */
+static inline int atomic_dec(atomic_int_t *var)
+{
+    int old;
+
+    do {
+        old = var->value;
+    } while (!atomic_cas(var, old, old - 1));
+
+    return old;
+}
+
+/**
+ * @brief Set an atomic variable to one.
+ *
+ * @param[inout]  var   Pointer to an atomic variable to update
+ *
+ * @return 1 if the old value was 0 and the variable was successfully updated
+ * @return 0 if the variable was already set
+ */
+static inline int atomic_set(atomic_int_t *var)
```

Comment:
Hm, I find the name misleading. I started commenting the brief description until I realized that you actually meant "set the variable to 1".
Also, where is the difference to ```atomic_set_return(var,1)```? Isn't this a convenience function to save that variable?